src/toil/lib/threading.py

```diff
@@ -105,42 +105,77 @@ def cpu_count() -> Any:
-            return total_machine_size
-
-        with open('/sys/fs/cgroup/cpu/cpu.cfs_period_us') as stream:
-            # Read the period in which we are allowed to burn the quota
-            period = int(stream.read())
+    try:
+        # See if we can fetch these and use them
+        quota: Optional[int] = None
+        period: Optional[int] = None
+
+        # CGroups v1 keeps quota and period separate
+        CGROUP1_QUOTA_FILE = '/sys/fs/cgroup/cpu/cpu.cfs_quota_us'
+        CGROUP1_PERIOD_FILE = '/sys/fs/cgroup/cpu/cpu.cfs_period_us'
+        # CGroups v2 keeps both in one file, space-separated, quota first
+        CGROUP2_COMBINED_FILE = '/sys/fs/cgroup/cpu.max'
+
+        if os.path.exists(CGROUP1_QUOTA_FILE) and os.path.exists(CGROUP1_PERIOD_FILE):
+            logger.debug('CPU quota and period available from cgroups v1')
+            with open(CGROUP1_QUOTA_FILE) as stream:
+                # Read the quota
+                quota = int(stream.read())
+
+            with open(CGROUP1_PERIOD_FILE) as stream:
+                # Read the period in which we are allowed to burn the quota
+                period = int(stream.read())
+        elif os.path.exists(CGROUP2_COMBINED_FILE):
+            logger.debug('CPU quota and period available from cgroups v2')
+            with open(CGROUP2_COMBINED_FILE) as stream:
+                # Read the quota and the period together
+                quota, period = [int(part) for part in stream.read().split(' ')]
+        else:
+            logger.debug('CPU quota/period not available from cgroups v1 or cgroups v2')
 
-        logger.debug('CPU quota period: %d', period)
+        if quota is not None and period is not None:
+            # We got a quota and a period.
+            logger.debug('CPU quota: %d period: %d', quota, period)
 
-        # The thread count is how many multiples of a wall clcok period we can burn in that period.
-        cgroup_size = int(math.ceil(float(quota)/float(period)))
+            if quota == -1:
+                # But the quota can be -1 for unset.
+                # Assume we can use the whole machine.
+                return total_machine_size
 
-        logger.debug('Cgroup size in cores: %d', cgroup_size)
+            # The thread count is how many multiples of a wall clock period we
+            # can burn in that period.
+            cgroup_size = int(math.ceil(float(quota)/float(period)))
 
+            logger.debug('Control group size in cores: %d', cgroup_size)
     except:
         # We can't actually read these cgroup fields. Maybe we are a mac or something.
         logger.debug('Could not inspect cgroup: %s', traceback.format_exc())
-        cgroup_size = float('inf') # type: ignore
+
+    # CPU affinity may limit the size
+    affinity_size: Union[float, int] = float('inf')
+    if hasattr(os, 'sched_getaffinity'):
+        try:
+            logger.debug('CPU affinity available')
+            affinity_size = len(os.sched_getaffinity(0))
```

Comment:
Consider using `len(psutil.Process().cpu_affinity()` which is supported for Linux, Windows, FreeBSD  (and we already use `psutil`)

https://psutil.readthedocs.io/en/latest/index.html#psutil.Process.cpu_affinity

Reply:
Ah, nevermind, the implementation is quite similar between the two

https://github.com/python/cpython/blob/0076ca48e9cd116bb9d283d014fe69a72c9f1f6f/Modules/posixmodule.c#L7229
https://github.com/giampaolo/psutil/blob/5b3016ba3d7429c827c992699f50c2b7ff29b023/psutil/_psutil_linux.c#L239